nova/cmd/manage.py

```diff
@@ -675,17 +675,34 @@ def null_instance_uuid_scan(self, delete=False):
           default=90, help='Days from today to begin purging')
     @args('--max-number', metavar='<number>', dest='max_number',
           help='Maximum number of instances to consider')
+    @args('--all-cells', action='store_true', dest='all_cells',
+          default=False, help='Run against all cell databases')
     def purge_deleted_instances(self, dry_run=False, older_than=90,
-                                max_number=None):
+                                max_number=None, all_cells=False):
         """Removes soft deleted instances data"""
         older_than = int(older_than)
         if max_number:
             max_number = int(max_number)
 
         admin_context = context.get_admin_context(read_deleted='yes')
-        instance_count = db.instances_purge_deleted(admin_context, dry_run,
-                                                    older_than,
-                                                    max_number)
+        instance_count = 0
+        if all_cells:
+            try:
+                cells = objects.CellMappingList.get_all(admin_context)
+            except db_exc.DBError:
+                print(_('Unable to get cell list from API DB. '
+                        'Is it configured?'))
+            for cell in cells:
+                with context.target_cell(admin_context, cell) as cctxt:
+                    instance_count += db.instances_purge_deleted(cctxt,
+                                                                 dry_run,
+                                                                 older_than,
+                                                                 max_number)
```

Comment:
Should `max_number` still be the same as given by the user even after we already deleted some instances in a previous cell? I think this is unexpected and we should pass `max_number - instance_count`.

Reply:
fixed